File: apps/ios/ScribbleFit/ScribbleFit/Features/Analytics/Data/Repository/AnalysisRepositoryImpl.swift
import Foundation
import SwiftData

/**
 * iOS implementation of AnalysisRepository using ScribbleFitDatabase (SwiftData).
 */
public final class AnalysisRepositoryImpl: AnalysisRepository {
    private let database: ScribbleFitDatabase
    private let jsonDecoder = JSONDecoder()
    private let jsonEncoder = JSONEncoder()
    
    public init(database: ScribbleFitDatabase = .shared) {
        self.database = database
    }
    
    public func getHomeSuggestion() async throws -> AnalysisSuggestion? {
        guard let jsonData = database.getInsightByKey(key: KEY_HOME_SUGGESTION)?.jsonData.data(using: .utf8) else {
            return nil
        }
        return try jsonDecoder.decode(AnalysisSuggestion.self, from: jsonData)
    }
    
    public func getSummary(period: SummaryPeriod) async throws -> AnalysisSummary? {
        let key = "\(KEY_SUMMARY_PREFIX)_\(period.rawValue)"
        guard let jsonData = database.getInsightByKey(key: key)?.jsonData.data(using: .utf8) else {
            return nil
        }
        return try jsonDecoder.decode(AnalysisSummary.self, from: jsonData)
    }
    
    public func getExerciseInsight(exerciseId: String) async throws -> ExerciseInsight? {
        let key = "\(KEY_EXERCISE_PREFIX)_\(exerciseId)"
        guard let jsonData = database.getInsightByKey(key: key)?.jsonData.data(using: .utf8) else {
            return nil
        }
        return try jsonDecoder.decode(ExerciseInsight.self, from: jsonData)
    }
    
    public func saveHomeSuggestion(_ suggestion: AnalysisSuggestion) async throws {
        let data = try jsonEncoder.encode(suggestion)
        if let jsonString = String(data: data, encoding: .utf8) {
            let entity = InsightsCache(
                key: KEY_HOME_SUGGESTION,
                jsonData: jsonString,
                createdAt: Date()
            )
            database.upsertInsight(entity)
        }
    }
    
    public func saveSummary(_ summary: AnalysisSummary) async throws {
        let key = "\(KEY_SUMMARY_PREFIX)_\(summary.period.rawValue)"
        let data = try jsonEncoder.encode(summary)
        if let jsonString = String(data: data, encoding: .utf8) {
            let entity = InsightsCache(
                key: key,
                jsonData: jsonString,
                createdAt: Date()
            )
            database.upsertInsight(entity)
        }
    }
    
    public func saveExerciseInsight(_ insight: ExerciseInsight) async throws {
        let key = "\(KEY_EXERCISE_PREFIX)_\(insight.exerciseId)"
        let data = try jsonEncoder.encode(insight)
        if let jsonString = String(data: data, encoding: .utf8) {
            let entity = InsightsCache(
                key: key,
                jsonData: jsonString,
                createdAt: Date()
            )
            database.upsertInsight(entity)
        }
    }
    
    public func clearOldInsights() async throws {
        database.clearInsights()
    }
    
    private let KEY_HOME_SUGGESTION = "home_suggestion"
    private let KEY_SUMMARY_PREFIX = "summary"
    private let KEY_EXERCISE_PREFIX = "exercise_insight"
}

File: apps/android/feature/ai/data/src/main/java/com/scribblefit/feature/ai/data/repository/AnalysisRepositoryImpl.kt
package com.scribblefit.feature.ai.data.repository

import com.scribblefit.core.database.dao.InsightsCacheDao
import com.scribblefit.core.database.model.InsightsCacheEntity
import com.scribblefit.feature.ai.data.mapper.*
import com.scribblefit.feature.ai.domain.model.*
import com.scribblefit.feature.ai.domain.repository.AnalysisRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AnalysisRepositoryImpl @Inject constructor(
    private val insightsCacheDao: InsightsCacheDao,
    private val json: Json
) : AnalysisRepository {

    override fun getHomeSuggestion(): Flow<AnalysisSuggestion?> {
        return insightsCacheDao.getInsightByKey(KEY_HOME_SUGGESTION).map { entity ->
            entity?.let { json.decodeFromString<SuggestionDto>(it.jsonData).toDomain() }
        }
    }

    override fun getSummary(period: SummaryPeriod): Flow<AnalysisSummary?> {
        val key = "${KEY_SUMMARY_PREFIX}_${period.name.lowercase()}"
        return insightsCacheDao.getInsightByKey(key).map { entity ->
            entity?.let { json.decodeFromString<SummaryDto>(it.jsonData).toDomain(period) }
        }
    }

    override fun getExerciseInsight(exerciseId: String): Flow<ExerciseInsight?> {
        val key = "${KEY_EXERCISE_PREFIX}_$exerciseId"
        return insightsCacheDao.getInsightByKey(key).map { entity ->
            entity?.let { json.decodeFromString<ExerciseInsightDto>(it.jsonData).toDomain(exerciseId) }
        }
    }

    override suspend fun saveHomeSuggestion(suggestion: AnalysisSuggestion) {
        val entity = InsightsCacheEntity(
            key = KEY_HOME_SUGGESTION,
            jsonData = json.encodeToString(suggestion.toDto()),
            createdAt = System.currentTimeMillis()
        )
        insightsCacheDao.upsertInsight(entity)
    }

    override suspend fun saveSummary(summary: AnalysisSummary) {
        val key = "${KEY_SUMMARY_PREFIX}_${summary.period.name.lowercase()}"
        val entity = InsightsCacheEntity(
            key = key,
            jsonData = json.encodeToString(summary.toDto()),
            createdAt = System.currentTimeMillis()
        )
        insightsCacheDao.upsertInsight(entity)
    }

    override suspend fun saveExerciseInsight(insight: ExerciseInsight) {
        val key = "${KEY_EXERCISE_PREFIX}_${insight.exerciseId}"
        val entity = InsightsCacheEntity(
            key = key,
            jsonData = json.encodeToString(insight.toDto()),
            createdAt = System.currentTimeMillis()
        )
        insightsCacheDao.upsertInsight(entity)
    }

    override suspend fun clearOldInsights() {
        insightsCacheDao.deleteAll()
    }

    companion object {
        private const val KEY_HOME_SUGGESTION = "home_suggestion"
        private const val KEY_SUMMARY_PREFIX = "summary"
        private const val KEY_EXERCISE_PREFIX = "exercise_insight"
    }
}
